csrc/dynamic_transform.cpp

```diff
@@ -308,20 +390,27 @@ void DynamicTransformConcretizer::mutate(TensorView* tv) {
             ". IterDomain was expected.");
       }
 
-      // If none of the output IDs is symbolic, nothing to concretize
-      if (std::all_of(
+      // If any output ID is symbolic, all output IDs should be symbolic
+      if (std::any_of(
               expr->outputs().begin(), expr->outputs().end(), [](Val* output) {
-                return output->as<IterDomain>()->getIterType() !=
+                return output->as<IterDomain>()->getIterType() ==
                     IterType::Symbolic;
               })) {
+        TORCH_INTERNAL_ASSERT(std::all_of(
+            expr->outputs().begin(), expr->outputs().end(), [](Val* output) {
+              return output->as<IterDomain>()->getIterType() ==
+                  IterType::Symbolic;
+            }));
+      } else if (std::all_of(
+                     expr->inputs().begin(),
+                     expr->inputs().end(),
+                     [](Val* output) {
+                       return output->as<IterDomain>()->getIterType() !=
+                           IterType::Symbolic;
+                     })) {
+        // If no inputs or outputs are symbolic, nothing to concretize
         continue;
       }
```

Comment:
@naoyam this change was necessary to fix failing test `DynamicTransform5_CUDA` which contains a dynamic reshape followed by a static pad (constant pad widths of 1, 1). The pad is inferred to be output itertype Iteration at definition, even though its input is symbolic due to the dynamic reshape. That means that concretization needs to handle a symbolic root domain with already-concrete rfactor domain. Previously, skipped mutation in this function if none of the ID expr outputs was symbolic. Now we skip if neither of inputs or outputs is symbolic instead. I preserved the check that if any output is symbolic then they all must be.

Reply:
I think this change is a hack and that the real problem is that `propagateFromProducerToConsumer` is not concretizing the root domain in this case (when the rfactor is concrete already). There may be an assumption in `PairwiseRootDomainMap` that `Symbolic` roots only appear with `Symbolic` rfactor domains. I am diving into this now.

Reply:
`propagateFromProducerToConsumer` works properly in this case, registering a mutation of the `Symbolic` root IDs in the pad. It's just that it was previously not processing all of the exprs between root and rfactor after that point. Now we process every expr between root and rfactor as long as it has a `Symbolic` input or output. I think this is correct.

Reply:
Thanks for the fix, makes sense. I think we could make this a little more simpler to follow. The conditions we would want to check would be:

1. All of inputs are non symbolic: Nothing to do. Just assert outputs are also non symbolic
2. All of outputs are non symbolic: Nothing to do. Just leave a comment that inputs can be symbolic with an example like the `pad` case
3. Continue otherwise

Reply:
You're right that is clearer. Changed the check and added a comment explaining the dynamic reshape->concrete pad case.

Reply:
It looks like `FusionSliceForNanoGPT3_CUDA` is a counterexample to the assertion that one symbolic output implies they all are symbolic. In that example, we input a tensor of shape `[16, 128, 3072]`, then slice it by `[:16, :128, :1024]`, followed by a reshape to size `[16, 128, 16, 64]`. The slice ranges and reshape sizes are all static, but since the input extents are symbolic, the slice is translated to a resize with symbolic pad widths, i.e. the resize is symbolic. **However**, since the reshape is static, the split factors are static, so the last dimension of the output rfactor is concrete:
```
Inputs:
  T0_g[ iS0{i0}, iS1{i2}, iS2{i3} ], float
Outputs:
  T7_g[ ?S39{16}, ?S40{128}, ?S43{16}rf, iS44{( ceilDiv(1024, 16) )}rf ], float 
```
This may be a bit of an artifact of static reshape, so if we are not planning to keep static reshape in the future then it may still be a valid condition. However, it seems like static reshapes would always have these static split factors that may appear in r-factor domains even when roots are pure symbolic.

Reply:
Now that I reread your comment I think I misinterpreted your assertions. It's not an all or nothing assertion on outputs. I'll fix that, then the nanogpt test should be fine. 

I think your point 1 may have an issue though. We may have a concrete input to resize that leads to symbolic output, whenever we have non constant pad widths. And vice versa, we can have symbolic inputs that lead to concrete outputs when we have large enough concrete pad widths.

So it seems to me that we only really know there's nothing to do when we have no symbolic inputs _or_ outputs in a single expression.

Reply:
I removed the outputs test and expanded the comment here.